test/extensions/filters/http/geoip/geoip_filter_test.cc

```diff
@@ -0,0 +1,278 @@
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/runtime/mocks.h"
+#include "test/mocks/stats/mocks.h"
+#include "test/test_common/registry.h"
+#include "test/test_common/test_runtime.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::DoAll;
+using testing::ReturnRef;
+using testing::SaveArg;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Geoip {
+namespace {
+
+class GeoipFilterTest : public testing::Test {
+public:
+  GeoipFilterTest()
+      : dummy_factory_(new DummyGeoipProviderFactory()), dummy_driver_(dummy_factory_->getDriver()),
+        empty_response_(absl::nullopt), dummy_city_("dummy_city"), dummy_country_("dummy_country"),
+        dummy_region_("dummy_region"), dummy_asn_("dummy_asn"), dummy_anon_response_("true") {}
+
+  void initializeFilter(const std::string& yaml) {
+    envoy::extensions::filters::http::geoip::v3::Geoip config;
+    TestUtility::loadFromYaml(yaml, config);
+    config_ = std::make_shared<GeoipFilterConfig>(config, "prefix.", stats_, runtime_);
+    filter_ = std::make_unique<GeoipFilter>(config_, dummy_driver_);
+    filter_->setDecoderFilterCallbacks(filter_callbacks_);
+  }
+
+  void initializeProviderFactory() {
+    TestScopedRuntime scoped_runtime;
+    scoped_runtime.mergeValues(
+        {{"envoy.reloadable_features.no_extension_lookup_by_name", "false"}});
+    Registry::InjectFactory<GeoipProviderFactory> registered(*dummy_factory_);
+    EXPECT_CALL(*dummy_driver_, getCity(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getCountry(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getRegion(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getAsn(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getIsAnonymous(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getIsAnonymousVpn(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getIsAnonymousTorExitNode(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getIsAnonymousHostingProvider(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+    EXPECT_CALL(*dummy_driver_, getIsAnonymousPublicProxy(_, _, _))
+        .WillRepeatedly(DoAll(SaveArg<0>(&captured_address_)));
+  }
+
+  void expectStats(const std::string& geo_header) {
+    EXPECT_CALL(stats_, counter(absl::StrCat("prefix.geoip.", geo_header, ".total")));
+    EXPECT_CALL(stats_, counter(absl::StrCat("prefix.geoip.", geo_header, ".hit")));
+  }
+
+  void expectHeader(const Http::TestRequestHeaderMapImpl& request_headers,
```

Comment:
Could you make this a `MATCHER_P`? (It would make it easier for the next person to find and move into a common place, if you don't want to put it somewhere common already, and you get better line numbers on failure with a proper EXPECT_THAT rather than a helper function.)

Also, right now you'd get an exception thrown on the EXPECT_EQ if the EXPECT_TRUE fails, and not very helpful output. As a MATCHER_P you could easily `*result_listener << "bla bla bla" << request_headers` on each type of failure.